package a4.atlas;

import "a4/io/A4.proto";

import "a4/atlas/Physics.proto";

import "a4/atlas/Trigger.proto";
import "a4/atlas/EventStreamInfo.proto";

import "a4/atlas/Photon.proto";
import "a4/atlas/Electron.proto";
import "a4/atlas/Muon.proto";
import "a4/atlas/Jet.proto";
import "a4/atlas/TrackHits.proto";

message TruthParticle
{
    optional LorentzVector p4 = 1;
    optional int32 charge = 2;
    optional int32 pdg_id = 3;
}

message Track
{
    optional LorentzVector p4 = 1;
    optional int32 charge = 2;
    optional int32 vertex_index = 3 [default=-1];
    optional TrackHits hits = 4;
    optional Perigee perigee = 5;
}

message RunLB
{
    optional int32 run = 1;
    optional int32 lumiblock = 2;
}

// Note: Many of the properties of this class will only be correct when 
//       aggregated by keying on `name`.

message InputFile
{
    optional string filename = 1;
    optional uint64 size = 2;
    
    optional string grl_string = 10;
}

message ProcessingStep
{
    optional string name = 1;

    optional uint64 input_events = 10 [default=0];
    
    // Bytes read to generate this data
    optional uint64 input_bytes_read = 20 [default=0];
    
    optional float walltime = 30 [default=0];
    optional float cputime = 31 [default=0];
    
    repeated InputFile input_files = 40;
}

message EventMetaData
{
    option (a4.io.major_version) = "v2011.11";
    option (a4.io.patch_version) = "1";

    optional bool simulation = 1 [(a4.io.merge)=MERGE_BLOCK_IF_DIFFERENT];
    repeated int32 mc_channel = 11 [(a4.io.merge)=MERGE_UNION];
    
    repeated Stream stream = 2 [(a4.io.merge)=MERGE_UNION];
    repeated string period = 3 [(a4.io.merge)=MERGE_UNION];
    repeated int32 run = 4 [(a4.io.merge)=MERGE_UNION];
    repeated RunLB lumiblock = 5 [(a4.io.merge)=MERGE_UNION];

    optional int32 event_count = 6 [(a4.io.merge)=MERGE_ADD];
    optional double sum_mc_weights = 7 [(a4.io.merge)=MERGE_ADD];
    optional double reweight_lumi = 8 [(a4.io.merge)=MERGE_BLOCK_IF_DIFFERENT];
    
    // Information about a run over the input files
    repeated ProcessingStep processing_steps = 9 [(a4.io.merge)=MERGE_UNION];
}

message Event
{
    optional int32 CLASS_ID = 1001;
    optional int32 run_number = 1;
    optional int32 event_number = 2;
    optional int32 lumi_block = 3;
    optional int32 bunch_crossing_id = 12;
    repeated Stream stream_tag = 13;

    optional int32 mc_channel_number = 14;
    optional double mc_event_weight = 4 [default = 1.0];
    optional float actual_interactions_per_crossing = 16;
    optional float average_interactions_per_crossing = 17;

    optional double scale_factor = 15 [default = 1.0];

    repeated Trigger triggers = 5;
    repeated Vertex vertices = 6;
    optional MissingEnergy met = 7;

    repeated Jet jets = 8;
    repeated Muon muons = 9;

    repeated Electron electrons = 10;
    repeated Photon photons = 11;
    repeated Track tracks = 18;

    repeated Muon muons_staco = 20;
    repeated Muon muons_muid = 21;

    optional MissingEnergy met_LocHadTopo_modified = 30;
    optional MissingEnergy met_RefFinal45 = 32;

    // Plain MET
    optional MissingEnergy met_RefFinal = 31;
    optional MissingEnergy met_MuonBoy = 33;
    optional MissingEnergy met_Muid = 34;
    optional MissingEnergy met_RefMuon_Track = 35;
    optional MissingEnergy met_CorrTopo = 36;
    optional MissingEnergy met_CellOut_em = 84;
    optional MissingEnergy met_CellOut_Eflow = 85;
    optional MissingEnergy met_CellOut_Eflow_Muid = 86;
    optional MissingEnergy met_Final = 37;
    optional MissingEnergy met_LocHadTopo = 38;
    optional MissingEnergy met_Muon = 39;
    optional MissingEnergy met_MuonMuid = 70;
    optional MissingEnergy met_RefEle = 71;
    optional MissingEnergy met_RefFinal_em = 72;
    optional MissingEnergy met_RefGamma = 73;
    optional MissingEnergy met_RefJet = 74;
    optional MissingEnergy met_RefMuon = 75;
    optional MissingEnergy met_RefMuon_Muid = 76;
    optional MissingEnergy met_RefMuon_Track_Muid = 77;
    optional MissingEnergy met_RefTau = 78;
    optional MissingEnergy met_SoftJets = 79;
    optional MissingEnergy met_Topo = 80;
    optional MissingEnergy met_Track = 81;
    optional MissingEnergy met_Truth = 82;
    optional MissingEnergy met_Truth_PileUp = 83;


    repeated Jet jets_antikt4h1topo = 40;
    repeated Jet jets_antikt4h1topoem = 41;

    optional int32 error_state_lar = 50;

    enum HFORDecision {
        None = 0;
        isBB = 1;
        isCC = 2;
        isC = 3;
        isLightFlavor = 4;
        kill = 5;
    }
    optional HFORDecision hfor_decision = 60;

    optional MissingEnergy truth_met_hard = 100;
    optional MissingEnergy truth_met_pileup = 101;
    repeated LorentzVector truth_neutrinos = 102;
    repeated Muon truth_electrons = 103;
    repeated Muon truth_muons = 104;
}
